### PR TITLE
feat(ui): Keyboard Navigation for Preference & Plugin Panels

### PR DESCRIPTION
**Pull request by Koranir**
While talking with RisingLeaf, I found out that the preferences and plugin panels didn't have keyboard navigation :(

I've added it in :)

### CODE REVIEW

File: source/PreferencesPanel.cpp [hunk 1]
```diff
@@ -155,25 +156,34 @@ bool PreferencesPanel::KeyDown(SDL_Keycode key, Uint16 mod, const Command &comma
...
 		hoverItem.clear();
+		selected = 0;
 	}
 	else if(key == 'o' && page == 'p')
 		Files::OpenUserPluginFolder();
 	else if((key == 'n' || key == SDLK_PAGEUP) && currentSettingsPage < SETTINGS_PAGE_COUNT - 1)
+	{
 		++currentSettingsPage;
+		selected = 0;
+		selectedItem = "";
```
Comment (quyykk) [inline source/PreferencesPanel.cpp:179]: Use `.clear()` instead

File: source/PreferencesPanel.cpp [hunk 2]
```diff
@@ -155,25 +156,34 @@ bool PreferencesPanel::KeyDown(SDL_Keycode key, Uint16 mod, const Command &comma
...
+	}
 	else if((key == 'r' || key == SDLK_PAGEDOWN) && currentSettingsPage > 0)
+	{
 		--currentSettingsPage;
+		selected = 0;
+		selectedItem = "";
```
Comment (quyykk) [inline source/PreferencesPanel.cpp:185]: here too

File: source/PreferencesPanel.cpp [hunk 3]
```diff
@@ -807,10 +745,23 @@ void PreferencesPanel::DrawSettings()
...
+			auto width = FontSet::Get(14).Width(setting);
+			table.SetHighlight(-120, width - 110);
+			table.DrawHighlight(back);
+		}
+
 		table.Draw(setting, isOn ? medium : dim);
 		table.Draw(text, isOn ? bright : medium);
 	}
+
+	if(selectedItem == "")
```
Comment (quyykk) [inline source/PreferencesPanel.cpp:763]: `.empty()`

File: source/PreferencesPanel.cpp [hunk 4]
```diff
@@ -42,6 +42,7 @@ this program. If not, see <https://www.gnu.org/licenses/>.
 #include <SDL2/SDL.h>
 
 #include <algorithm>
+#include <string>
```
Comment (RisingLeaf) [inline source/PreferencesPanel.cpp:45]: already included in the header

File: source/PreferencesPanel.cpp [hunk 5]
```diff
@@ -807,10 +745,23 @@ void PreferencesPanel::DrawSettings()
...
+			auto width = FontSet::Get(14).Width(setting);
+			table.SetHighlight(-120, width - 110);
+			table.DrawHighlight(back);
+		}
+
 		table.Draw(setting, isOn ? medium : dim);
 		table.Draw(text, isOn ? bright : medium);
 	}
+
+	if(selectedItem.empty())
```
Comment (RisingLeaf) [inline source/PreferencesPanel.cpp:763]: comment pls

**Review by RisingLeaf — COMMENTED** [review]
Nice

File: source/PreferencesPanel.h [hunk 6]
```diff
@@ -56,6 +56,12 @@ class PreferencesPanel : public Panel {
 
 	void Exit();
 
+	void HandleSettingsString(const std::string &str, Point cursorPosition);
```
Comment (RisingLeaf) [inline source/PreferencesPanel.h:59]:
```suggestion
	void HandleSettingsString(const std::string &str, const Point &cursorPosition);
```
is this something we do? idk

File: source/PreferencesPanel.cpp [hunk 7]
```diff
@@ -932,3 +883,169 @@ void PreferencesPanel::Exit()
...
+	{
+	case 'c':
+		if(static_cast<unsigned>(selected + 1) < zones.size())
+			selected++;
+		break;
+	case 's':
+		if(static_cast<unsigned>(selected) >= prefZones.size())
+			selected = prefZones.size() - 1;
+		if(static_cast<unsigned>(selected + 1) < prefZones.size())
+			selected++;
```
Comment (RisingLeaf) [inline source/PreferencesPanel.cpp:1018]:
```suggestion
		selected = max(prefZones.size() - 1, selected + 1);
```
is this the same?

File: source/PreferencesPanel.h (hunk 6 above)
Comment (Koranir) [reply source/PreferencesPanel.h:59]: I mutate the value within the function, so I'd rather copy the value here.

File: source/PreferencesPanel.cpp [hunk 8]
```diff
@@ -489,7 +414,8 @@ void PreferencesPanel::DrawControls()
 			bool isHovering = (index == hover && !isEditing);
 			if(!isHovering && index == selected)
 			{
-				table.SetHighlight(-120, 54);
+				auto textWidth = FontSet::Get(14).Width(command.Description());
+				table.SetHighlight(-120, textWidth - 110);
```
Comment (Amazinite) [inline source/PreferencesPanel.cpp:418]: :pray:
That annoyed me. Thank you for fixing it.
image
image

File: source/PreferencesPanel.cpp [hunk 9]
```diff
@@ -807,10 +744,24 @@ void PreferencesPanel::DrawSettings()
 			text = isOn ? "on" : "off";
 
 		if(setting == hoverItem)
+		{
+			table.SetHighlight(-120, 120);
 			table.DrawHighlight(back);
+		}
```
Comment (Amazinite) [inline source/PreferencesPanel.cpp:750]: The mouse hovering is behaving funky in the settings. I sent a video on Discord showing what I mean.

File: source/PreferencesPanel.cpp [hunk 10]
```diff
@@ -539,6 +465,17 @@ void PreferencesPanel::DrawSettings()
...
+	if(hoverItem != oldHoverItem)
+	{
+		latestItem = hoverItem;
+		oldHoverItem = latestItem;
+	}
+	if(selectedItem != oldSelectedItem)
+	{
+		latestItem = selectedItem;
+		oldSelectedItem = latestItem;
+	}
```
[the remark this answers is not in the document]
Comment (Koranir) [reply source/PreferencesPanel.cpp]: Ah yeah, that was a holdover from copying over code. DrawControls uses latest to figure out which controls should be disabled when 'x' is pressed, but DrawSettings doesn't have that :).

File: source/PreferencesPanel.h [hunk 11]
```diff
@@ -69,7 +75,10 @@ class PreferencesPanel : public Panel {
 
 	Point hoverPoint;
 	int hoverCount = 0;
+	std::string selectedItem;
 	std::string hoverItem;
+	std::string oldSelectedItem;
+	std::string oldHoverItem;
```
Comment (Amazinite) [inline source/PreferencesPanel.h:81]: Well now these are unused too.

Comment (Koranir) [reply source/PreferencesPanel.h]: I feel like C++ should give warnings for unused variables :P

**Pull request merged**